### DISCUSSION TITLE
MicroProfile Rest Client - Multiple multipart files

### DISCUSSION THREAD

**Discussion by vinimdev** [quarkusio/quarkus#39751] [Q&A] [answered]
Guys, I am using Quarkus 3.8.3 and the MicroProfile Rest Client to make HTTP requests, but I am having problems with uploading multiple files.

I am using the client in this way:

```java
@POST
@Consumes(MediaType.MULTIPART_FORM_DATA)
Response postRespbuilderWithFiles(@HeaderParam(AUTHORIZATION) String authorization,
                                  @HeaderParam(CANAL_ATENDIMENTO) String canalAtendimento,
                                  @HeaderParam(EMAIL) String email,
                                  @HeaderParam(USUARIO) String usuario,
                                  @RestQuery MultivaluedMap<String, ?> parametros,
                                  QuarkusMultipartForm files);
```

and assembling the files in this way:

```java
public QuarkusMultipartForm buildClientMultipartForm(List<FileUpload> files) {
        QuarkusMultipartForm multiPartForm = new QuarkusMultipartForm();
        files.forEach(fu -> multiPartForm.binaryFileUpload("files", fu.fileName(), fu.filePath().toString(), fu.contentType()));
        return multiPartForm;
}
```

This has worked correctly when there's only one file, but when there are multiple files, I receive from my other receiving application (Spring or Quarkus) that the body is malformed:

```posh
Caused by: org.springframework.web.multipart.MultipartException: Failed to parse multipart servlet request; nested exception is java.lang.RuntimeException: java.io.IOException: io.undertow.util.MalformedMessageException
	at org.springframework.web.multipart.support.StandardMultipartHttpServletRequest.handleParseFailure(StandardMultipartHttpServletRequest.java:127)
	at org.springframework.web.multipart.support.StandardMultipartHttpServletRequest.parseRequest(StandardMultipartHttpServletRequest.java:115)
	at org.springframework.web.multipart.support.StandardMultipartHttpServletRequest.<init>(StandardMultipartHttpServletRequest.java:88)
	at org.springframework.web.multipart.support.StandardServletMultipartResolver.resolveMultipart(StandardServletMultipartResolver.java:122)
	at org.springframework.web.servlet.DispatcherServlet.checkMultipart(DispatcherServlet.java:1209)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1043)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:964)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	... 51 common frames omitted
```

How can I solve this problem with multiple files?

[a comment by an automated account is not shown]

**Comment by geoand** [comment]
Something is definitely not correct as your application seems to be a Spring application judging from the stacktrace

**Reply by vinimdev** [reply]
The error log is from the receiving application; the originating application (Quarkus) does not give an error when sending the files. It's the receiver that cannot interpret the files when there is more than one file. I tested with a Quarkus application as the receiver, and the error is of the same type.

**Comment by vinimdev** [comment]
Here's how it arrives at the receiving application:

Request with more than one file:
<img width="1336" alt="Multiple files" src="https://github.com/quarkusio/quarkus/assets/60160909/180b533c-4cc5-4073-a2af-603168544e45">

Request with just one file:
<img width="952" alt="One file" src="https://github.com/quarkusio/quarkus/assets/60160909/346707ab-d53b-4680-9821-581bab1e989c">

**Comment by geoand** [comment]
I'd need two sample application (client and server) in order to be able to tell you more

**Reply by vinimdev** [reply]
I pointed the receiver to a copy of my original app (Quarkus) and the error is the same, here is a screenshot of the error and how the controller looks, it doesn't even reach the controller:

<img width="1304" alt="Captura de Tela 2024-03-27 às 16 15 52" src="https://github.com/quarkusio/quarkus/assets/60160909/e42e91b9-3f3e-4cba-9795-2d5cacc31591">

**Reply by vinimdev** [reply]
with one file, it arrives normally:
<img width="1350" alt="Captura de Tela 2024-03-27 às 16 17 54" src="https://github.com/quarkusio/quarkus/assets/60160909/6a99a370-43ac-4a88-90c3-9c3c4eb671f4">

**Reply by vinimdev** [reply]
This discussion, in which he provides two applications, one in Quarkus and the other in Spring, is exactly my problem. It works perfectly with just one file, but with more than one file, it presents the problem.

https://github.com/quarkusio/quarkus/discussions/36780

**Reply by geoand** [reply]
I tried those and could not reproduce the problem you mention.

Please attach a spring and a quarkus application here that I can use to easily reproduce the exact problem you are seeing.

**Reply by vinimdev** [reply]
I discovered that the error in Spring only occurs when I'm using Undertow.

```xml
<dependency>
  <groupId>org.springframework.boot</groupId>
  <artifactId>spring-boot-starter-web</artifactId>
  <exclusions>
    <exclusion>
      <groupId>org.springframework.boot</groupId>
      <artifactId>spring-boot-starter-tomcat</artifactId>
    </exclusion>
  </exclusions>
</dependency>
<dependency>
  <groupId>org.springframework.boot</groupId>
  <artifactId>spring-boot-starter-undertow</artifactId>
</dependency>
```

**Reply by vinimdev** [reply]
But from Quarkus to Quarkus, the problem persists.

**Reply by vinimdev** [reply]
Here's an example of the Quarkus app, what I did was create a replica and call Quarkus itself and it gives the mentioned error.

And for Spring, the error occurs when using Undertow.

https://github.com/vinimdev/quarkus-example

**Reply by geoand** [reply]
Thanks, I'll soon have a look

**Reply by vinimdev** [reply]
Example spring with undertow (error)

https://github.com/vinimdev/spring-undertow-exaple

**Reply by geoand** [reply, marked as answer]
Take a look at https://github.com/quarkusio/quarkus/pull/39770 for an explanation of the problem.

For now what you can do is configure 

```
quarkus.rest-client.multipart-post-encoder-mode=HTML5
```

but when that PR is merged and included in a release, you will be able to do:

```
quarkus.rest-client.spring-multipart.multipart-post-encoder-mode=HTML5
```

**Reply by rsvoboda** [reply]
HTMT5 => HTML5 @geoand ?

**Reply by geoand** [reply]
Good point!

**Reply by rsvoboda** [reply]
I have updated https://github.com/quarkusio/quarkus/discussions/39751#discussioncomment-8940253 comment

**Comment by geoand** [comment]
https://github.com/quarkusio/quarkus/discussions/39751#discussioncomment-8940253 is the final answer :)

**Reply by vinimdev** [reply]
@geoand  The error remained even after adding: HTMT5 or HTML5

**Reply by geoand** [reply]
I tried it and it worked for me on your app

**Reply by vinimdev** [reply]
Could it be because I'm using new builder? Is there a way to set this property programmatically?

```java
public <T> T createProduceClient(String url, Class<T> client, Long connectTimeout, Long readTimeout) {
      return RestClientBuilder.newBuilder()
              .baseUri(URI.create(url))
              .register(new ResponseSizeLimitFilter(responseSizeLimit))
                .register(new HeaderRequestFilter(userContext))
                .connectTimeout(connectTimeout)
                .readTimeout(readTimeout)
                .property("microprofile.rest.client.disable.default.mapper", true)
                .build(client);
    }
```

**Reply by vinimdev** [reply]
Got it this way @geoand:

Thanks for the help, mate!

```java
.property(QuarkusRestClientProperties.MULTIPART_ENCODER_MODE, PausableHttpPostRequestEncoder.EncoderMode
                        .valueOf("HTML5"))
```

**Reply by geoand** [reply]
Right, I was about to say you can do that, but I will improve my [PR](https://github.com/quarkusio/quarkus/pull/39770) so you don't have to do that.

**Reply by geoand** [reply]
Once that is in you will be able to do `builder.multipartPostEncoderMode("HTML5")`
